Akka.Cluster.Metrics extension implementation

IgorFedchenko opened the pull request:
This will close #2070 , and maybe some other related issues (like #2069)

The goal of this PR is to port [Akka.Cluster.Metrics](https://github.com/akka/akka/tree/master/akka-cluster-metrics) into Akka.NET. Maybe this will span into several pull requests, will see.

I will keep this in draft state until will have working specs for this (that is, once this will be ready for final reviews/merge).

I will start from implementing the core functions and metric collectors, and then will add adaptive load balancing routers.

IgorFedchenko commented:
Completed port of the extension code base.

Need to add/port specs and verify it is working.

Aaronontheweb commented on src/contrib/cluster/Akka.Cluster.Metrics/Serialization/StandardMetrics.cs:
```diff
@@ -0,0 +1,204 @@
+    public static class StandardMetrics
+    {
+        // Constants for the heap related Metric names
+        public const string HeapMemoryUsed = "heap-memory-used";
```
Some ways of collecting this information in .NET:

1. The garbage collector: https://github.com/petabridge/NBench/blob/dev/src/NBench/Collection/Memory/GcTotalMemoryCollector.cs#L34
2. Performance Counters (Windows only)
3. Calling into native APIs via P/Invoke - would need different implementations for Windows vs. POSIX

Aaronontheweb replied:
There might also be some new .NET Core instrumentation that is cross-platform. That's been on the to-do list for CoreCLR for some time now but I think they've made some progress against that since 2.2+

IgorFedchenko replied:
Hmm, indeed - I will need to address this once will work on .NET metrics collector here, in this PR. I totally agree that we should look for something like _total allocated memory_, and will update collector's name (and metric's name as well) once will get clear understanding of what we are collecting at the end.

IgorFedchenko replied:
Thanks for hints about research directions, by the way - and did not know anything about JVM heap :)

Aaronontheweb commented on src/contrib/cluster/Akka.Cluster.Metrics/reference.conf:
```diff
@@ -0,0 +1,111 @@
+  }
+  # Globally unique metrics extension serializer identifier.
+  serialization-identifiers {
+    "akka.cluster.metrics.protobuf.MessageSerializer" = 10
```
This namespace is wrong - needs to be a CLR fully-qualified type name. Also, is the number 10 taken already? I don't think so since this functionality was originally built into Akka.Cluster, but it'd be good to check (although I guess that will come up when you start porting the multi-node tests.)

IgorFedchenko replied:
You know, that's funny - 7/8/9 are taken, 11/12/13 are taken, but 10 is not (searching in configurations by `serialization-identifiers` string). So will keep it like that, and will see what multi-node tests will show.
Thanks for type name hint - fixed it

Aaronontheweb replied:
Sounds like this is resolved then.

Aaronontheweb commented on src/core/Akka/Util/Option.cs:
```diff
@@ -77,5 +89,15 @@ public override int GetHashCode()
+        /// Applies specified action to this option, if there is a value
+        /// </summary>
+        /// <param name="action"></param>
+        public void ForEach(Action<T> action)
```
I'd rename this to be more idiomatic in C#, rather than borrow the Scala API

IgorFedchenko replied:
Oh yes, I spent some small time thinking about more idiomatic name, but nothing got to my mind and I left it as it is... Renamed it to `OnSuccess` now - do you think it is more intuitive/idiomatic?

Aaronontheweb replied:
Yep, that's better.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Helpers/DynamicAccess.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <summary>
+    /// DynamicAccess
+    /// </summary>
+    internal static class DynamicAccess
```
Good idea. Also I will add `Try` class to `Akka.Utils`, since it is used by `DynamicAccess` and also may be used somewhere else.

Aaronontheweb replied:
Might also be worth getting rid of `DynamicAccess` altogether and just creating some `Activator` extension methods - more idiomatic that way. What do you think?

IgorFedchenko replied:
Mmm what do you mean by `Activator` extension methods? Obviously is can not add custom static method like `Activator.CreateInstanceFor<TSomeInterface>(string typeName)`. Either use `(new Activator()).CreateInstanceFor`, or create another static class like `ActivatorHelper` - but I named it `DynamicAccess` to just simplify porting scala code in the future (sometimes one needs to check if some akka's class is already implemented, and search by name helps in such cases).

We can also add extension method for string, like `typeName.CreateInstanceAs<TInterface>(params object[] args)`, but... It would be hard to find for other developers, and not obvious.

So since  `DynamicAccess.CreateInstanceFor` wraps getting type by name, then instantiating with `Activator`, and handling exceptions (it returns `Try<TResult>` instance which may contain error info), it might be useful when porting things in future. But I can rename it to something more intuitive - do you have any suggestions? I agree that nobody would look for `DynamicAccess` class if only he is not porting `dynamicAccess` class from scala at the moment.

Aaronontheweb replied:
@IgorFedchenko let's make `DynamicAccess` `internal` and then keep it as-is - that way if we decided to clean up the API to make it more idiomatic with .NET in the future we can do that without worrying about user-facing issues.

IgorFedchenko replied:
To keep it in `Akka.Utils` to make reusable in different modules marked it with `[InternalApi]` and added `INTERNAL USAGE` text into class summary.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/ClusterMetrics.cs:
```diff
@@ -0,0 +1,124 @@
+        /// Creates new <see cref="ClusterMetrics"/> for given actor system
+        /// </summary>
+        /// <param name="system"></param>
+        public ClusterMetrics(ExtendedActorSystem  system)
```
Made this `internal` since it is used in `ClusterMetricsExtensionProvider`.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/ClusterMetricsCollector.cs:
```diff
@@ -0,0 +1,209 @@
+
+        private void SendGossip(Akka.Actor.Address address, MetricsGossipEnvelope envelope)
+        {
+            Context.ActorSelection(Self.Path.ToStringWithAddress(address)).Tell(envelope);
```
Maybe this is to ensure `At-Least-Once-Delivery` of gossip message? If I remember right, `Self.Tell` may not be delivered if `Self` will be restarted? And `ActorSelection` allows newly created instance to handle message? At least, this is the first of two usual reasons to use `ActorSelection` (where second is remote connection).

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/ClusterMetricsSupervisor.cs:
```diff
@@ -0,0 +1,96 @@
+                case ClusterMetricsSupervisorMetadata.CollectionStartMessage start:
+                    Context.GetChildren().ForEach(child => Context.Stop(child));
+                    _collectorInstance++;
+                    Context.ActorOf(Props.Create<ClusterMetricsSupervisor>(), CollectorName);
```
Ah, good point - no, of course not, it should be `ClusterMetricsCollector`.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/ClusterMetricsSupervisor.cs:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /// <summary>
+    /// ClusterMetricsSupervisorMetadata
```
Internal - left it public for writing specs, but sure need to add `InternalApi` attribute and corresponding line in summary.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Configuration/ClusterMetricsSettings.cs:
```diff
@@ -0,0 +1,62 @@
+        // Extension.
+        public string MetricsDispatcher => _config.GetString("dispatcher");
+        public TimeSpan PeriodicTasksInitialDelay => _config.GetTimeSpan("periodic-tasks-initial-delay");
+        public string NativeLibraryExtractFolder => _config.GetString("native-library-extract-folder");
```
If you are about `NativeLibraryExtractFolder` setting - yes, turned out that is was used only for some built-in collector (Sigar, and btw they have bindings for .net, so maybe sometime we will need to use it too: https://github.com/hyperic/sigar).
But removed it for now.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Helpers/TryNative.cs:
```diff
@@ -0,0 +1,132 @@
+    /// <summary>
+    /// Wraps possible exceptions into <see cref="Try{T}"/> result
+    /// </summary>
+    internal static class TryNative
```
Ok, got it.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/MetricsCollectorBuilder.cs:
```diff
@@ -0,0 +1,56 @@
+    /// Metrics collector instantiation priority order:
+    /// 1. Provided custom collector
+    /// 2. Internal <see cref="DummyCollector"/>
+    /// </summary>
```
I think I will get back to this once the module will be working and will make separate PR with documentation for it (including docs for site). What do you thing?

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Routing/ClusterMetricsRouting.cs:
```diff
@@ -0,0 +1,429 @@
+            // The current weighted routees, if any. Weights are produced by the metricsSelector
+            // via the metricsListener Actor. It's only updated by the actor, but accessed from
+            // the threads of the sender()s.
+            _weightedRouteesRef = new AtomicReference<Tuple<ImmutableArray<Routee>, IImmutableSet<NodeMetrics>, Option<WeightedRoutees>>>(
```
I wish I could - but `AtomicReference` requires it's type argument to be `class`, which is not the case for `ValueTuple`. 
For this reason I had to stick with `Tuple` class in some other parts of code base once was migrating it to `ValueTuple`.
P.S. And in turn `AtomicReference` can not use `struct` types since `Interlocked.CompareExchange` requires a reference.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Serialization/ClusterMetricsMessageSerializer.cs:
```diff
@@ -0,0 +1,370 @@
+        {
+            switch (obj)
+            {
+                case MetricsGossipEnvelope m: return Compress(MetricsGossipEnvelopeToProto(m)); // TODO: Add compression here
```
Yes, I just left it there to not forget about this idea when/if metrics communication will become slow.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Serialization/EWMA.cs:
```diff
@@ -0,0 +1,97 @@
+            ///             the sampled value resulting from the previous smoothing iteration.
+            ///             This value is always used as the previous EWMA to calculate the new EWMA.
+            /// </summary>
+            public sealed partial class EWMA 
```
This class is `partial`, and it may require more efforts to use separate `EWMA` struct, because it is used by other `partial` classes... So if it is not a significant improvement, maybe better to keep it as it is - do you agree?

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Serialization/Proto/ClusterMetricsMessages.proto:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Defines a remote address.
+ */
+message Address {
```
It turned out that yes, we have `AddressData` message defined in `protobuf/ContainerFormats.proto` (and generated as `Akka.Remote.Serialization.Proto.Msg` package).
But for some reason generated c# `AddressData` class is `internal` there, and I can not reuse it.
Should I regenerate proto messages with `public` modifier, or just have my own `Address` message defined as we have now?

IgorFedchenko replied:
Ah, almost forgot about this. Thanks for the tip, fixed now.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Routing/MetricSelectors.cs:
```diff
@@ -0,0 +1,239 @@
+        public override IImmutableDictionary<Actor.Address, decimal> Capacity(IImmutableSet<NodeMetrics> nodeMetrics)
+        {
+            return nodeMetrics
+                .Select(StandardMetrics.HeapMemory.Unapply)
```
There is only `Unapply`, which is basically converting data object to tuple containing some of it's fields. Or to the `Option<ValueTuple>` to be more accurate, since requested metrics may be missing.
Have just renamed it to `Decompose` to reflect the semantics.

IgorFedchenko commented:
@Aaronontheweb big thanks for a deep review! Updated code to reflect changes we discussed - and left questions about few moments. This was very helpful, and I made few notes for the future.

In the meantime, added some specs and made them passing (which required some changes to ported code), and working on next specs.

IgorFedchenko commented:
Now all specs (except spec for metric collector, which is not implemented yet) are added and ported code is tested/fixed.

The next, and last step here is to implement .NET metrics collector, and add tests for it.

IgorFedchenko commented:
@Aaronontheweb Speaking about measuring memory usage in .NET Core (the thread you started [here](https://github.com/akkadotnet/akka.net/pull/4126#discussion_r361715588) , looks like some nice work was done by .net core team in the past year - mostly by adding/fixing `Process` class API. I.e. [this](https://github.com/dotnet/corefx/issues/23449) PR was closed few months ago, and allows us to use `Process.GetCurrentProcess().PrivateMemorySize64` under Linux as well for example. `WorkingSet64` is also fixed.

My current idea is to use `WorkingSet64` property as a metric to collect for routing. `PrivateMemorySize64` will be collected as well, and `GC.GetTotalMemory` as an additional info that can be used by any node in cluster.

IgorFedchenko commented:
Added `DefaultCollector` to collect memory and cpu metrics. Had to remove `SystemAverageLoad` metric collector to simplify things for now.

Here are the issues we need to overcome right now:

- To support load-balanced routing, we need to get memory capacity - something in `[0, 1]` range where 0 is "no memory available" and 1 is "all memory is available, pick me!". While `Process.WorkingSet64` is somewhat close to "total physical memory in RAM actually used by process" (but still maybe need to use another source), we need to know the total memory available for process on machine. That is, ideally,  we need `available` memory - something not allocated by any process yet, and `(available - used) / available` would be what we need. 
But in .NET Core there is no cross-platform way to get free memory, or even just available hardware RAM installed. Right now I am using `Process.VirtualMemorySize64` to get something that is limiting available memory for the process as `available` value, but this may be fine for routing (since only relative values between nodes matters), but not great as an absolute value to share with other nodes since neither `Process.VirtualMemorySize64` or `Process.VirtualMemorySize64 - Process.WorkingSet64` reflect the memory load well.
For now using some native APIs looks most clear way to go.

- Mostly the same about CPU. To avoid using `PerformanceCounters` and have cross-platform code, I am measuring `Process.TotalProcessorTime` for current process and comparing it to total time elapsed (close to what is described [in this article](https://medium.com/@jackwild/getting-cpu-usage-in-net-core-7ef825831b8b)). The problem here is that for routing ideally would be using all processes CPU load - that is, take all processes into account, not only current process.
Right now, I am just using current process CPU load for routing - i.e. 3% on first node is always considered better then 5% on other node, even if on first node there are +80% CPU load from other processes.
To get all processes CPU load, I could also track all processes via `Process.GetProcesses()` - and I tried that, but this is working slow (because it is loading all info about all processes, catching access violations if any - lots of overhead).
So to improve that, if we want to consider all processes load, native APIs are again most promising way.

So, @Aaronontheweb , there is no problem to collect as many metrics as we want (from GC, from Process, and some OS-specific metrics), but the problem is to get values for routing. 
Should we complicate code with native API invocations and environment switches, or do you see another way to move on? Or just keep things as they are implemented now and publish `beta` package version to get some feedback?

Aaronontheweb commented:
@IgorFedchenko 

So the crux of the issue is that getting access to machine-level metrics is going to require native calls, by the sound of it. Ran into this issue with Akka.NET 1.1 back before .NET Core was released. I was hoping it had been addressed by now... Sounds like not.

What'd I'd recommend we do:

1. Ship a version of Cluster.Metrics that works as-is and uses "good enough" weighted routing to get the job done. What _mostly_ matters is the busy-ness of each node relative to the others, not the absolute value. The only downside is that for Akka.Cluster.Metrics scenarios outside of weighted routing (i.e. subscribing to metrics gossip for auto-scaling purposes) these metrics won't be accurate enough to support those. However, first principles first: :shipit: _ship it_.
2. In the background, develop a separate library outside of Akka.NET that can use native Windows / POSIX p/invoke calls down into the correct OS APIs to gather these metrics. That's going to be more portable than PerformanceCounters (which, I should add, have regular problems from a data corruption and permissions standpoint - seen it time and time again with NBench). We can use an adapter pattern of sorts to change out the P/Invoke implementation from Windows to POSIX to retrieve the same types of data.

Some resources you might find helpful in that regard:

1. https://docs.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke - P/Invoke programming guide for C#
2. https://www.pinvoke.net/ - helpful encyclopedia for p/invoke-ing into a large number of Windows APIs
3. https://developers.redhat.com/blog/2016/09/14/pinvoke-in-net-core-rhel/ - example from RedHat of P/Invoking into RHEL (Red Hat Enterprise Linux) APIs; I'm sure there are other resources for doing this on Linux with .NET Core but this one is pretty good.

Aaronontheweb commented:
By the way, I think the implementation you've put together thus far and the explanations for it are very good given the API / scope creep constraints you're under. Nice work. Is this PR ready for final review now?

IgorFedchenko commented:
@Aaronontheweb 

> By the way, I think the implementation you've put together thus far and the explanations for it are very good given the API / scope creep constraints you're under. Nice work. Is this PR ready for final review now?

Thanks for good words!
I am still working to make MNTR tests work, and after that will let you know once PR is ready to be merged (from my point of view).

> What'd I'd recommend we do:

Agree with you. Also, if I will put description for metrics like I did for `MemoryUsed` (so that people know that it is `Process.WorkingSet64`) even absolute values may be useful for users. The only think missing will be absolute values of "capacity" - available memory and CPU on particular node. But indeed this may be not a big issue - at least, we can resolve it later with some external library.

So yes, new `Akka.Cluster.Metrics` module is almost there - I hope to get it ready for final review on Thursday/Friday.

IgorFedchenko commented:
Almost working now, but for some reason the last part of one of MNTR specs is failing. [This one](https://github.com/IgorFedchenko/akka.net/blob/a390887309953be0999e6847611d97410cb6e76f/src/contrib/cluster/Akka.Cluster.Metrics.Tests.MultiNode/ClustetMetricsRoutingSpec.cs#L297-L300).

Another part of the specs showed that weighted router is working, collecting messages from all nodes, and even performs weighted selection. But it seems like there are some issues when enabling clustering on router.

IgorFedchenko commented:
@Aaronontheweb All right, MNTR tests are passing locally now. Will see what CI say about that.

Also, I have added `-Dmultinode.clear-output=1` option for MNTR, because it is very annoying to perform logs cleanup each time when you are debugging and looking for fresh logs each time. When set to something greater then 0 (which is the default), `TestResults/multinode` directory will be deleted before tests start.
I know, this is not directly related to this PR, and I can create separate one for this minor change, but I really needed this feature here.

One more thing that is not directly related to this PR but requiring fix was `AwaitAssertAsync` method from my recent PR about async methods in TestKit. When using code like this:
```csharp
await AwaitAssertAsync(async () =>
 {
    var r = await OperationAsync();
    r.Should().BeTrue();
});
```
there was missing overload `AwaitAssertAsync(Task<Func> funcAsync)`, and code above was using `AwaitAssertAsync(Action action)` - and of course did not awaiting for inner async completion. So added this missing overload.

IgorFedchenko commented:
@Aaronontheweb Now should be ready for review/merge.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics.Tests.MultiNode/Sample/StatsSampleSpec.cs:
```diff
@@ -0,0 +1,137 @@
+                
+                cluster.Unsubscribe(TestActor);
+                
+                TestConductor.Enter("all-up");
```
Indeed

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/Collectors/DefaultCollector.cs:
```diff
@@ -0,0 +1,133 @@
+                {
+                    // Memory
+                    NodeMetrics.Types.Metric.Create(StandardMetrics.MemoryUsed, process.PrivateMemorySize64).Value,
+                    // VirtualMemorySize64 is not best idea here...
```
@Aaronontheweb Created detailed issue about this - tried to give some understanding of current implementation state, and direction to move on.

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/StandardMetrics.cs:
```diff
@@ -0,0 +1,193 @@
+        /// <summary>
+        /// Total memory allocated to the currently running process (<see cref="Process.PrivateMemorySize64"/>)
+        /// </summary>
+        public const string MemoryUsed = "MemoryUsed";
```
Sure! Thanks

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics/reference.conf:
```diff
@@ -0,0 +1,107 @@
+# This is the reference config file that contains all the default settings.
+# Make your edits in your application.conf in order to override these settings.
+
+# Sigar provisioning:
```
@Aaronontheweb I will, but actually seems like this library has [bindings for .NET](https://github.com/hyperic/sigar/tree/master/bindings/dotnet). I knew that when started porting this modure, but do not remember the reason why I have excluded this from port. Maybe to start with more simple implementation, that will not require bundled binaries... Anyway, asked about Sigar here: https://github.com/akkadotnet/akka.net/issues/4142#issuecomment-574310269.

Aaronontheweb replied:
Saw that issue - good idea. We can address that in next pass.

IgorFedchenko commented on src/core/Akka.Cluster/Routing/ClusterRoutingConfig.cs:
```diff
@@ -253,21 +253,18 @@ internal static string UseRoleOption(string role)
-        /// <param name="config">TBD</param>
-        /// <returns>TBD</returns>
-        internal static int GetMaxTotalNrOfInstances(Config config)
+        public static int GetMaxTotalNrOfInstances(Config config)
```
`InternalsVisibleTo` should be just fine, I do not think this method is going to be useful for external usage. Thanks!

IgorFedchenko commented on src/core/Akka.MultiNodeTestRunner/Program.cs:
```diff
@@ -132,13 +132,16 @@ static void Main(string[] args)
             var platform = CommandLine.GetPropertyOrDefault("multinode.platform", "net");
             var reporter = CommandLine.GetPropertyOrDefault("multinode.reporter", "console");
+            
+            var clearOutputDirectory = CommandLine.GetInt32OrDefault("multinode.clear-output", 0);
```
Glad to know :)

IgorFedchenko commented on src/core/Akka.Remote.TestKit/MultiNodeSpec.cs:
```diff
@@ -506,6 +506,16 @@ public void RunOn(Action thunk, params RoleName[] nodes)
+        /// Execute the given block of code only on the given nodes (names according
+        /// to the `roleMap`).
+        /// </summary>
+        public async Task RunOnAsync(Func<Task> thunkAsync, params RoleName[] nodes)
```
Well - this particular overload I added because `thunk` node action was containing some of `Akka.TestKit` methods, which I was using in new `async` variation. So to use `AwaitAssertAsync` and others we need to have this `RunOnAsync` overload.
This is the most popular blocking methods. Another blocking methods are about event filtering (i.e. `ExpectMsg`), and they turned out to be using `BlockingCollection` internally - so no way to easily convert them to `async` methods without changing the event capturing implementation.
We may add `async` overloads to all (I mean to less popular as well) test methods that use blocking API anyway, for example `EnterBarrier` and other. Well, `EnterBarrier` may block bunch of threads during execution actually (several nodes, several parallel tests, every node is waiting for others on the barrier...) Should I create issue for this?

Aaronontheweb replied:
Yes, create an issue with making an async overload for EnterBarrier. If we decide to implement it we will want to convert all of our MNTR specs to return `Task` as well.

IgorFedchenko replied:
Done: https://github.com/akkadotnet/akka.net/issues/4146

IgorFedchenko commented:
@Aaronontheweb Made updates, and also have some comment threads unresolved.

Aaronontheweb commented:
@IgorFedchenko looks like all of the recent test failures are specific to Akka.Cluster.Metrics.

IgorFedchenko commented:
> @IgorFedchenko looks like all of the recent test failures are specific to Akka.Cluster.Metrics.

Oh yes, one minor typo in `reference.conf` file and it all crashes... Thanks :)

Aaronontheweb commented:
Need API approvals and one more routing fix

Sent from my iPhone

> On Jan 15, 2020, at 10:43 AM, Igor Fedchenko <notifications@github.com> wrote:
> 
> ﻿
> @IgorFedchenko commented on this pull request.
> 
> In src/contrib/cluster/Akka.Cluster.Metrics/Serialization/Proto/ClusterMetricsMessages.proto:
> 
> > +    Number number = 2;
> +    EWMA ewma = 3;
> +  }
> +  int32 addressIndex = 1;
> +  int64 timestamp = 2;
> +  repeated Metric metrics = 3;
> +}
> +
> +/****************************************
> + * Common Datatypes and Messages
> + ****************************************/
> +
> +/**
> + * Defines a remote address.
> + */
> +message Address {
> Ah, almost forgot about this. Thanks for the tip, fixed now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

IgorFedchenko commented:
Added API approval, but do not understand why used memory may be `0`. Actually, [here](https://github.com/dotnet/corefx/issues/23449) people say that `Process.PrivateMemorySize64` is returning 0 under linux, and looks like this issue is fixed only in `netcoreapp3.1`, and I missed this, but... Let me run CI again for now.

Aaronontheweb commented:
That explains why `Akka.Cluster.Metrics.Tests.MetricValuesSpec.NodeMetrics_MetricValues_should_extract_expected_MetricValue_types_for_load_balancing` is failing on Linux.

@IgorFedchenko this is hacky, but we should probably check to see which version of the .NET Core runtime we're running on while on Linux and display a warning message from inside Akka.Cluster.Metrics that as a result of this bug, you will not get accurate memory metrics while running on Linux unless you upgrade to .NET Core 3.1 or later. That will save us some support ticket trouble down the road.

For this specific test, I'd perform the same check and simply not run the test if platform != .NET Core 3.1 or later on Linux. How does that sound?

IgorFedchenko commented:
> That explains why `Akka.Cluster.Metrics.Tests.MetricValuesSpec.NodeMetrics_MetricValues_should_extract_expected_MetricValue_types_for_load_balancing` is failing on Linux.
> 
> @IgorFedchenko this is hacky, but we should probably check to see which version of the .NET Core runtime we're running on while on Linux and display a warning message from inside Akka.Cluster.Metrics that as a result of this bug, you will not get accurate memory metrics while running on Linux unless you upgrade to .NET Core 3.1 or later. That will save us some support ticket trouble down the road.
> 
> For this specific test, I'd perform the same check and simply not run the test if platform != .NET Core 3.1 or later on Linux. How does that sound?

Alternatively, I could use `GC.GetTotalMemory` instead - this would simplify code and protect users from versioning issues. We are going to improve metrics collection anyway at https://github.com/akkadotnet/akka.net/issues/4142, and `GC.GetTotalMemory` should not give much difference with `Process.PrivateMemorySize64`, at least in most scenarios, especially from routing point of view.

So can change metric value here, and have it working well. What do you think? @Aaronontheweb

Aaronontheweb commented:
@IgorFedchenko fantastic idea - let's do it.

IgorFedchenko commented:
> @IgorFedchenko fantastic idea - let's do it.

Done.

Aaronontheweb commented:
From the most recent build:

```
Akka.Cluster.Metrics.Tests.MetricValuesSpec.NodeMetrics_MetricValues_should_extract_expected_MetricValue_types_for_load_balancing

Expected value to be between 0.0 and 1.0 because CPU process usage should be between 0% and 100%, but found 1.4242444383254682.
```

Is that a math error of some kind? That'd be 142% of CPU by this measure, right?

IgorFedchenko commented:
> Is that a math error of some kind? That'd be 142% of CPU by this measure, right?

This may be a linux issue of `process.TotalProcessorTime`, since this failure is reproduced second time only under linux. As per routing, values greater then 1 are also handled in scala implementation (and in ours), as 0% capacity. So maybe this is something people were getting sometimes as well.

But actually found possible reason why sometimes it may have such big values - maybe the issue is in tracking time elapsed between measurements... Have just fixed this.

Aaronontheweb commented:
@IgorFedchenko 

Akka.Cluster.Metrics.Tests.MetricValuesSpec.NodeMetrics_MetricValues_should_extract_expected_MetricValue_types_for_load_balancing
Expected value to be between 0.0 and 1.0 because CPU process usage should be between 0% and 100%, but found 1.0438517964271876.

Have that spec failing across all three framwork / platform combinations.

IgorFedchenko commented:
And of course it does not want to be reproducible on local machine... 

All right, making CPU tracking in the non-blocking way turned out to be little bit more tricky. I think now I put everything on the shelves - let's see what CI thinks about it.

Aaronontheweb commented:
@IgorFedchenko 

> And of course it does not want to be reproducible on local machine...

Oh man... I just realized that the issue might be that the way those metrics get reported to you from within a Hypervisor might be skewed possibly. I don't _think_ that should happen but once you start tapping into system APIs at this level weird shit starts to happen.

IgorFedchenko commented:
@Aaronontheweb 

All right, so another run without new module-related failures, and next run (after merging branch) got this CPU >100% issue, only on Windows at this time. 

I am not sure that I can do anything with this staying with current CPU metrics loading implementation. While this is the only way to get cross-platform CPU usage at .net core (at least this is what I could find after some long research). 
Actually, fixes I have done so far were mostly for edge cases I thought could be the reason of CI failures. Like "if calling `Process.GetCurrentProcess` takes long enough time, and there are almost no more heavy processes running at the same time, this could give us not fully correct value". There is no much ways for improvements that I can see - except changing the CPU metrics collection approach completely.

The next step I would do here is to move to some native API's (using Sigar, or direct P/Invoke, etc) - all that stuff we are targeting at #4142. 

As per this PR, CPU load reported by `DefaultCollector` is definitely correlating with real CPU load for currently running process anyway, so it's values should be more or less useful, and give acceptable results on smart routing.

So publishing this package, maybe with some pre-release versioning, sounds like a good idea to me. This could also help us gathering feedback about other issues, like bad values from Hypervisor or due to other virtual environments specific - the kind of information it is hard/time consuming to catch during development.

So, let me know what you think. I will appreciate any ideas about improving this PR as well.

P.S. Ah, almost forgot - for now I have just limited reported CPU load value to `100%`, to make 100+% be reported as a valid "maximum possible load" value.

Aaronontheweb commented:
If those tests are flaky because of issues with the accuracy of real CPU measurement, just disable them. Better to not have them running then

IgorFedchenko commented:
Disabled, but no the one that was failing due to CPU measurement (because that one I have just fixed with 100% limit), but the one that was failing sometimes and was about measuring performance.

CPU 100% limit I want to keep there because this is generally fine for users to expect values less then 100% in that metric. So people may be surprised and even have bugs when value is growing above the limit. So cutting it on 100% for now.

Aaronontheweb commented:
@IgorFedchenko have this failing in the MNTR: https://dev.azure.com/dotnet/Akka.NET/_build/results?buildId=27573&view=logs&j=f203efd3-114c-5599-f6b4-0fa25ffc9760&t=b10ae0da-ff8b-5c6e-60d8-f20e659bbc17&l=8341

```
[Node #1(node-1)][ERROR][1/23/2020 8:51:21 PM][Thread 0018][akka://ClusterMetricsEnabledSpec/system/log2-TestEventListener] The type initializer for 'Akka.Cluster.Metrics.Serialization.ClusterMetricsMessagesReflection' threw an exception.
```

IgorFedchenko commented:
> @IgorFedchenko have this failing in the MNTR: https://dev.azure.com/dotnet/Akka.NET/_build/results?buildId=27573&view=logs&j=f203efd3-114c-5599-f6b4-0fa25ffc9760&t=b10ae0da-ff8b-5c6e-60d8-f20e659bbc17&l=8341
> 
> ```
> [Node #1(node-1)][ERROR][1/23/2020 8:51:21 PM][Thread 0018][akka://ClusterMetricsEnabledSpec/system/log2-TestEventListener] The type initializer for 'Akka.Cluster.Metrics.Serialization.ClusterMetricsMessagesReflection' threw an exception.
> ```

@Aaronontheweb Can not reproduce this locally, and can not find anything about this in Azure logs - url you provided shows "Build not found" error, and didn't find such failures in last CI runs.

But `ClusterMetricsMessagesReflection` is a static class generated by Protobuf, that is performing some basic setup, so... Do not have an idea of why it could be failing (seems like it's static constructor throws some exception, that is wrapped into `TypeInitializationException`).

With that said, we could get real exception in our logs from `TypeInitializationException`'s `InnerException` property - but looks like `Akka.NodeTestRunner` does not report inner exceptions for non-`AggregateException` types. Opened an issue about that #4162 , and added closing PR. So after merging that one, the next time we will get real issue logged.

Aaronontheweb commented:
@IgorFedchenko merged in the latest changes - going to re-run the test suite now.

My guess is that it's a mismatched manifest in the serializer - sending one thing but trying to read another.

IgorFedchenko commented:
> My guess is that it's a mismatched manifest in the serializer - sending one thing but trying to read another.

Turned out that `Akka.Remote.Serialization.Proto.Msg.AddressData` class we are reusing from `Akka.Remote` project was internal, and related property in some protobuf messages was internal, unlike other properties. So most likely this was an issue (error said that defined property was not found). Fixed it, making `Akka.Remote.Serialization.Proto.Msg.AddressData` public but with `[InternalApi]` attribute.

Also, disabled one MNTR spec phase where one of the nodes was allocating memory and we were asserting that this node will receive less messages. This is actually not true in MNTR tests, because all nodes share common system resources and most likely to have same memory capacity from the router point of view. Weird spec, skipped it for now.

And tried to fix another racing spec with increasing timeout of waiting for cluster members to be up.

Actually, as I can see, there were quite a few errors, so @Aaronontheweb thanks again for pointing me out to the logs output.

Aaronontheweb commented:
> Fixed it, making Akka.Remote.Serialization.Proto.Msg.AddressData public but with [InternalApi] attribute.

That'll break the next time we generate the protobuf APIs. In that case I'd just create your own `AddressData` object in the protobuf definition... Which, ironically, I told you not to do in one of the earlier reviews... My bad 😨

IgorFedchenko commented:
@Aaronontheweb Updated this shared `AddressData` stuff, now `Akka.Remote`'s version is internal as it was. Actually, that's me who ought to check this visibility issues before reusing existing code.

I have a question about one last MNTR test failing, do not have an idea why it could pass locally but fail on CI.
Basically, here is what I am doing in test (that's the very first line of the spec):
```c#
AwaitClusterUp(roles: _config.NodeList.ToArray());
```

And it fails, with something like this on each node:
```
[Node #4(node-4)]
[Node #4(node-4)][Node4:node-4][FAIL] Akka.Cluster.Metrics.Tests.MultiNode.ClusterMetricsEnabledSpec.ClusterMetrics_Should_work_when_enabled
[Node #4(node-4)][Node4:node-4][FAIL-EXCEPTION] Type: Akka.TestKit.Xunit2.Internals.AkkaEqualException
[Node #4(node-4)]--> [Node4:node-4][FAIL-EXCEPTION] Message: Assert.Equal() Failure
[Node #4(node-4)]Expected: 5
[Node #4(node-4)]Actual:   0
[Node #4(node-4)]--> [Node4:node-4][FAIL-EXCEPTION] StackTrace:    at Akka.TestKit.Xunit2.XunitAssertions.AssertEqual[T](T expected, T actual, String format, Object[] args) in D:\a\1\s\src\contrib\testkits\Akka.TestKit.Xunit2\XunitAssertions.cs:line 67
[Node #4(node-4)]   at Akka.Cluster.TestKit.MultiNodeClusterSpec.<>c__DisplayClass26_0.<AwaitMembersUp>b__2() in D:\a\1\s\src\core\Akka.Cluster.TestKit\MultiNodeClusterSpec.cs:line 367
[Node #4(node-4)]   at Akka.TestKit.TestKitBase.AwaitAssert(Action assertion, Nullable`1 duration, Nullable`1 interval) in D:\a\1\s\src\core\Akka.TestKit\TestKitBase_AwaitAssert.cs:line 46
[Node #4(node-4)]   at Akka.Cluster.TestKit.MultiNodeClusterSpec.<>c__DisplayClass26_0.<AwaitMembersUp>b__0() in D:\a\1\s\src\core\Akka.Cluster.TestKit\MultiNodeClusterSpec.cs:line 368
[Node #4(node-4)]   at Akka.TestKit.TestKitBase.<>c__DisplayClass151_0.<Within>b__0() in D:\a\1\s\src\core\Akka.TestKit\TestKitBase_Within.cs:line 57
[Node #4(node-4)]   at Akka.TestKit.TestKitBase.Within[T](TimeSpan min, TimeSpan max, Func`1 function, String hint, Nullable`1 epsilonValue) in D:\a\1\s\src\core\Akka.TestKit\TestKitBase_Within.cs:line 134
[Node #4(node-4)]   at Akka.TestKit.TestKitBase.Within(TimeSpan min, TimeSpan max, Action action, String hint, Nullable`1 epsilonValue) in D:\a\1\s\src\core\Akka.TestKit\TestKitBase_Within.cs:line 58
[Node #4(node-4)]   at Akka.TestKit.TestKitBase.Within(TimeSpan max, Action action, Nullable`1 epsilonValue) in D:\a\1\s\src\core\Akka.TestKit\TestKitBase_Within.cs:line 33
[Node #4(node-4)]   at Akka.Cluster.TestKit.MultiNodeClusterSpec.AwaitMembersUp(Int32 numbersOfMembers, ImmutableHashSet`1 canNotBePartOfMemberRing, Nullable`1 timeout) in D:\a\1\s\src\core\Akka.Cluster.TestKit\MultiNodeClusterSpec.cs:line 374
[Node #4(node-4)]   at Akka.Cluster.TestKit.MultiNodeClusterSpec.AwaitClusterUp(RoleName[] roles) in D:\a\1\s\src\core\Akka.Cluster.TestKit\MultiNodeClusterSpec.cs:line 271
[Node #4(node-4)]   at Akka.TestKit.TestKitBase.AwaitAssertAsync(Action assertion, Nullable`1 duration, Nullable`1 interval) in D:\a\1\s\src\core\Akka.TestKit\TestKitBase_AwaitAssert.cs:line 85
[Node #4(node-4)]   at Akka.Cluster.Metrics.Tests.MultiNode.ClusterMetricsEnabledSpec.Should_collect_and_publish_metrics_and_gossip_them_around_the_node_ring() in D:\a\1\s\src\contrib\cluster\Akka.Cluster.Metrics.Tests.MultiNode\ClusterMetricsExtensionSpec.cs:line 99
[Node #4(node-4)]   at Akka.Cluster.Metrics.Tests.MultiNode.ClusterMetricsEnabledSpec.ClusterMetrics_Should_work_when_enabled() in D:\a\1\s\src\contrib\cluster\Akka.Cluster.Metrics.Tests.MultiNode\ClusterMetricsExtensionSpec.cs:line 88
[Node #4(node-4)]   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
[Node #4(node-4)]--- End of stack trace from previous location where exception was thrown ---
[Node #4(node-4)]   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
[Node #4(node-4)]   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

But before this failure, there is a weird error:
```
[Node #1(node-1)][ERROR][1/25/2020 1:28:25 PM][Thread 0018][akka://ClusterMetricsEnabledSpec/user/controller/barriers] [[akka://ClusterMetricsEnabledSpec/user/controller/%5B%3A%3A1%5D%3A1782%3A1782-server5#1634863782]] tried to enter 'node-2-started' while we were waiting for 'node-1-started'
[Node #1(node-1)]Cause: Akka.Remote.TestKit.BarrierCoordinator+WrongBarrierException: [[akka://ClusterMetricsEnabledSpec/user/controller/%5B%3A%3A1%5D%3A1782%3A1782-server5#1634863782]] tried to enter 'node-2-started' while we were waiting for 'node-1-started'
[Node #1(node-1)]   at Akka.Remote.TestKit.BarrierCoordinator.<>c__DisplayClass14_6.<InitFSM>b__15(EnterBarrier barrier) in D:\a\1\s\src\core\Akka.Remote.TestKit\BarrierCoordinator.cs:line 576
[Node #1(node-1)]   at Akka.Case.With[TMessage](Action`1 action) in D:\a\1\s\src\core\Akka\PatternMatch.cs:line 107
[Node #1(node-1)]   at Akka.Remote.TestKit.BarrierCoordinator.<InitFSM>b__14_2(Event`1 event) in D:\a\1\s\src\core\Akka.Remote.TestKit\BarrierCoordinator.cs:line 573
[Node #1(node-1)]   at Akka.Actor.FSM`2.ProcessEvent(Event`1 fsmEvent, Object source) in D:\a\1\s\src\core\Akka\Actor\FSM.cs:line 1210
[Node #1(node-1)]   at Akka.Actor.FSM`2.Receive(Object message) in D:\a\1\s\src\core\Akka\Actor\FSM.cs:line 1115
[Node #1(node-1)]   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message) in D:\a\1\s\src\core\Akka\Actor\ActorBase.cs:line 158
[Node #1(node-1)]   at Akka.Actor.ActorCell.ReceiveMessage(Object message) in D:\a\1\s\src\core\Akka\Actor\ActorCell.DefaultMessages.cs:line 177
[Node #1(node-1)]   at Akka.Actor.ActorCell.Invoke(Envelope envelope) in D:\a\1\s\src\core\Akka\Actor\ActorCell.DefaultMessages.cs:line 83
```

So how is it possible that one node is entering next barrier while others are waiting for the first one?
Here is a `AwaitClusterUp` code in our test suite that contains this barriers:
```c#
public void AwaitClusterUp(params RoleName[] roles)
{
	// make sure that the node-to-join is started before other join
	RunOn(StartClusterNode, roles.First());

	EnterBarrier(roles.First().Name + "-started");
	if (roles.Skip(1).Contains(Myself)) Cluster.Join(GetAddress(roles.First()));

	if (roles.Contains(Myself))
	{
		AwaitMembersUp(roles.Length);
	}
	EnterBarrier(roles.Select(r => r.Name).Aggregate((a, b) => a + "-" + b) + "-joined");
}
```
Do you have an idea why this could happen?

Also, I would not say that this particular multi node test is very useful - it checks that extension is working at all, while other tests are also using this extension (so checking it is working), so I could skip the test. But just not like the idea that something is not working and I do not understand why. But not sure that this is related to `Akka.Cluster.Metrics` extension at all.

Aaronontheweb commented:
> Do you have an idea why this could happen?

lol.... I think I do. A fun but common distributed programming bug. That code assumes that the `roles` collection has all nodes stored in the same order.... There's no guarantee that this ordering is consistent across all nodes, as far as I can remember. Let me look into this.

Aaronontheweb commented on src/contrib/cluster/Akka.Cluster.Metrics.Tests.MultiNode/ClusterMetricsExtensionSpec.cs:
```diff
@@ -0,0 +1,164 @@
+        public readonly RoleName Node4;
+        public readonly RoleName Node5;
+
+        public IImmutableSet<RoleName> NodeList => ImmutableHashSet.Create(Node1, Node2, Node3, Node4, Node5);
```
Ah, this is your bug.... The hashset may not list these in a consistent order.

IgorFedchenko replied:
Ahh.... Exactly... They are unordered here... Missed this... Thanks!!

IgorFedchenko commented on src/contrib/cluster/Akka.Cluster.Metrics.Tests.MultiNode/ClusterMetricsExtensionSpec.cs:
```diff
@@ -93,7 +91,7 @@ private async Task Should_collect_and_publish_metrics_and_gossip_them_around_the
             await AwaitAssertAsync(() =>
             {
-                AwaitClusterUp(roles: _config.NodeList.ToArray());
+                AwaitClusterUp(Roles.ToArray());
```
@Aaronontheweb Could you give me a hint for the future: what is the difference here? Isn't `Roles` just same list of roles stored internally?
I see that I am using this `Roles` in another spec that passes without issues, and maybe this will work here too - but why?

Aaronontheweb replied:
Roles is something that gets populated automatically by the `MultiNodeSpec` base class - the content of the list is ordered and populated by the base classes themselves. The issue was that the `HashSet<T>` you were using could change the roles slightly since I think the `HashCode`s were different on a process by process basis.

IgorFedchenko replied:
You are right, or course, set does not guarantee same ordering... Great catch, @Aaronontheweb , thanks a lot!

Aaronontheweb commented:
@IgorFedchenko looks like we need API Approvals again here: https://dev.azure.com/dotnet/Akka.NET/_build/results?buildId=27752&view=ms.vss-test-web.build-test-results-tab&runId=1085134&resultId=100383&paneView=debug - this is due to the changes in Akka.Remote.

IgorFedchenko commented:
> @IgorFedchenko looks like we need API Approvals again here: https://dev.azure.com/dotnet/Akka.NET/_build/results?buildId=27752&view=ms.vss-test-web.build-test-results-tab&runId=1085134&resultId=100383&paneView=debug - this is due to the changes in Akka.Remote.

Done
